### PR TITLE
Implement search in descriptions

### PR DESCRIPTION
**Pull request by FluorescentHallucinogen**
graphql-voyager-search

### CODE REVIEW

File: src/components/doc-explorer/OtherSearchResults.tsx [hunk 1]
```diff
@@ -1,16 +1,17 @@
 import * as React from 'react';
 
 import { isMatch, highlightTerm } from '../../utils';
+import Markdown from '../utils/Markdown';
 
-interface SearchResultsProps {
+interface OtherSearchResultsProps {
```
Comment (IvanGoncharov) [inline src/components/doc-explorer/OtherSearchResults.tsx:6]: Please separate this and other fixes improvements refactoring into separate PRs.
It will make it easier to review this one.

File: src/components/doc-explorer/Argument.css [hunk 2]
```diff
@@ -65,11 +65,4 @@
   & .arg-description > p {
     margin: 0;
   }
-
-  & .arg-description:before {
-    display: block;
-    content: "#";
-    float: left;
-    margin-right: var(--spacing-unit);
-  }
```
Comment (IvanGoncharov) [inline src/components/doc-explorer/Argument.css:74]: Can you please explain why you removed it?

File: src/utils/highlight.tsx [hunk 3]
```diff
@@ -11,11 +11,7 @@ export function highlightTerm(content: string, term: string) {
 
   // Apply highlight to all odd elements
   for (var i = 1, length = result.length; i < length; i += 2) {
-    result[i] = (
-      <span key={i} style={{ backgroundColor: '#ffff03' }}>
-        {result[i]}
-      </span>
-    );
+    result[i] = <mark key={i}>{result[i]}</mark>;
```
Comment (IvanGoncharov) [inline src/utils/highlight.tsx:14]: please extract into separate PR

File: src/introspection/introspection.ts [hunk 4]
```diff
@@ -138,7 +138,7 @@ function markRelayTypes(schema: SimplifiedIntrospectionWithIds): void {
         return;
       }
 
-      const edgesType = connectionType.fields.edges.type
+      const edgesType = connectionType.fields.edges.type;
```
Comment (IvanGoncharov) [inline src/introspection/introspection.ts:141]: separate PR

File: src/components/utils/Markdown.tsx [hunk 5]
```diff
@@ -12,19 +13,31 @@ export default class Markdown extends React.Component<MarkdownProps> {
 
   constructor(props) {
     super(props);
-    this.renderer = new HtmlRenderer({ safe: true });
```
Comment (IvanGoncharov) [inline src/components/utils/Markdown.tsx:15]: Why you remove `safe`?

File: src/components/utils/Markdown.tsx [hunk 6]
```diff
@@ -12,19 +13,31 @@ export default class Markdown extends React.Component<MarkdownProps> {
...
+      this.props.text !== nextProps.text || this.props.termToHighlight !== nextProps.termToHighlight
+    );
+  }
+
+  highlightTermInMarkdown(content: string, term: string) {
+    if (!term) {
+      return content;
+    }
+
+    return content.replace(new RegExp(term, 'gi'), match => `<mark>${match}</mark>`);
```
Comment (IvanGoncharov) [inline src/components/utils/Markdown.tsx:31]: It doesn't use `this` please extract it as a normal function.

File: src/components/doc-explorer/TypeList.tsx [hunk 7]
```diff
@@ -37,15 +37,15 @@ export default class TypeList extends React.Component<TypeListProps> {
...
+      if (!(isMatch(type.name, filter) || isMatch(type.description, filter))) {
         return null;
       }
 
       return (
         <div key={type.id} className={classNames('typelist-item', className)}>
           <TypeLink type={type} onClick={onTypeLink} filter={filter} />
           <FocusTypeButton onClick={() => onFocusType(type)} />
-          <Description className="-doc-type" text={type.description} />
+          <Description className="-doc-type" text={type.description} termToHighlight={filter} />
```
Comment (IvanGoncharov) [inline src/components/doc-explorer/TypeList.tsx:48]: Please use a consistent naming either rename it to `filter` or create separate PR that renames `filter` to something else in current codebase.

File: src/components/doc-explorer/TypeDetails.tsx [hunk 8]
```diff
@@ -61,22 +62,3 @@ export default class TypeDetails extends React.Component<TypeDetailsProps> {
     );
   }
 }
-
```
Comment (IvanGoncharov) [inline src/components/doc-explorer/TypeDetails.tsx:64]: It's ok to use internal components right?
Why you moved it into a separate file?
Anyway, such changes should be done in separate PR.

File: src/components/doc-explorer/DocExplorer.tsx [hunk 9]
```diff
@@ -64,12 +64,12 @@ export default class DocExplorer extends React.Component<DocExplorerProps> {
...
+          placeholder={`Search ${name}...`}
+          value={currentNav.searchValue}
+          onSearch={this.handleSearch}
+        />
         <div className="scroll-area">
-          <SearchBox
-            placeholder={`Search ${name}...`}
-            value={currentNav.searchValue}
-            onSearch={this.handleSearch}
-          />
```
Comment (FluorescentHallucinogen) [inline src/components/doc-explorer/DocExplorer.tsx:72]: Moved into separate pull request #158.

File: src/components/doc-explorer/OtherSearchResults.tsx (hunk 1 above)
Comment (FluorescentHallucinogen) [reply src/components/doc-explorer/OtherSearchResults.tsx:6]: Done. Moved into separate pull request #159.

File: src/utils/highlight.tsx (hunk 3 above)
Comment (FluorescentHallucinogen) [reply src/utils/highlight.tsx:14]: Done. Moved into separate pull request #159.

File: src/introspection/introspection.ts (hunk 4 above)
Comment (FluorescentHallucinogen) [reply src/introspection/introspection.ts:141]: Done. Moved into separate pull request #159.

File: src/components/doc-explorer/TypeDetails.tsx (hunk 8 above)
Comment (FluorescentHallucinogen) [reply src/components/doc-explorer/TypeDetails.tsx:64]: Done. Moved into separate pull request #159.

Comment (FluorescentHallucinogen) [reply src/components/doc-explorer/TypeDetails.tsx:64]: I've spent some time to find the `EnumValue` declaration. Other classes are declared in the separate files with the same name as a class. That's why I've refactored it into separate file.

File: src/components/doc-explorer/Argument.css (hunk 2 above)
Comment (FluorescentHallucinogen) [reply src/components/doc-explorer/Argument.css:74]: I've removed it for consistency with other descriptions (type descriptions and field descriptions have no leading `#`). I've spent some time to figure out why this symbol appears (I thought that markdown is broken).

File: src/components/utils/Markdown.tsx (hunk 5 above)
Comment (FluorescentHallucinogen) [reply src/components/utils/Markdown.tsx:15]: I've removed it because I use `<mark>` HTML tag for highlighting the search term inside Markdown code. This is allowed by Markdown spec. I've tried to wrap the search term in the final HTML string, but this is not a trivial task. E.g. if the search term is `p`, it breaks the HTML code: `<<mark>p</mark>>Some descri<mark>p</mark>tion text<<mark>p</mark>>`.

File: src/components/doc-explorer/TypeList.tsx (hunk 7 above)
Comment (FluorescentHallucinogen) [reply src/components/doc-explorer/TypeList.tsx:48]: I've used the `termToHighlight` as a prop name for `<Markdown>` and `<Description>` (that wraps `<Markdown>`) components, because *inside* these components it's the only meaningful name (generally speaking, the value that should be highlighted doesn't have to be a search term). Outside of these components I've preserved the `filter` and `searchValues` that are not consistent throughout the codebase (in each case it depends of the context).

File: src/components/utils/Markdown.tsx (hunk 6 above)
Comment (FluorescentHallucinogen) [reply src/components/utils/Markdown.tsx]: Done.

**Comment by IvanGoncharov** [comment]
@FluorescentHallucinogen Can you please rebase?

**Comment by FluorescentHallucinogen** [comment]
@IvanGoncharov Done.

**Comment by orefalo** [comment]
Well, I find this to be a pretty cool addition

**Comment by LunaticMuch** [comment]
I closing this as it got stale.
